lib/connection.js

```diff
@@ -45,6 +46,11 @@ class Connection {
   }
 
   connect () {
+    // If the connection has already been established, don't do anything
+    if (this._connection.threadId) {
+      return Promise.resolve()
+    }
```

Comment:
This check is really just to avoid having to know if we're already connected inside `Connection.withConnection()`. The only reason that `Connection.withConnection()` would have an already connected connection is because `Pool.withConnection()` will get a new connection (`PoolConnection`) and pass it to `Connection.withConnection()` and pool connections are already connected when we get them.